### PR TITLE
NoSuchBeanDefinitionException in spring-bean's old version

### PR DESCRIPTION
**Pull request by Patrick0308**
fixed issue #806 NoSuchBeanDefinitionException. `@Autowired(required = false)` doesn't work in spring-bean's old version.

### CODE REVIEW

**Comment by bruno-garcia** [comment]
Thanks for raising this @Patrick0308 .
Do you mind adding a test so we can make sure the bootstrap works alright?

**Comment by Patrick0308** [comment]
Do you mean adding a spring boot test? @bruno-garcia

**Comment by bruno-garcia** [comment]
We have some tests here 
https://github.com/getsentry/sentry-java/blob/master/sentry-spring-boot-starter/src/test/java/io/sentry/spring/autoconfigure/SentryAutoConfigurationTest.java
I’m wondering is something can be added that exemplifies the change (repro the error before your change and fixes after).

**Comment by Patrick0308** [comment]
@bruno-garcia I don't know how to test in different spring-bean version. May be I use previous version in tests?

**Comment by bruno-garcia** [comment]
The versions we're testing are here:

https://github.com/getsentry/sentry-java/blob/28a9b1c1fa02bf5c30c8efbfd263a156ff417677/sentry-spring-boot-starter/pom.xml#L15

Did you experience errors with this one or a different?

**Comment by Patrick0308** [comment]
In a 1.5.3.RELEASE version of spring-boot-starter-parent,  an error occurred on application startup. In `pom.xml` , we use 1.5.17.RELEASE version now.

**Comment by bruno-garcia** [comment]
`1.5.17.RELEASE` is what our tests are targeting so it should work without this code change.
Have you tested this code change on ` 1.5.3.RELEASE` before upgrading to ` 1.5.17.RELEASE` to validate the fix works?
I'm just unsure if it's worth changing this since there are not tests to validate the behavior before/after. Since we test against ` 1.5.17.RELEASE` it can be enough to say we don't support version ` 1.5.3.RELEASE`, for example.

**Comment by Patrick0308** [comment]
Yeah, I've tested in my application on `1.5.3.RELEASE`. I suggest to support 1.5.x versions, because a lot of company use one of these versions.

**Comment by metlos** [comment]
Would you please then try reverting our `spring-boot-dependencies` version to say `1.5.0.RELEASE` and write a test for the changed behavior?

Just an FYI, `1.5.0.RELEASE` was released on 30 Jan 2017, `1.5.3.RELEASE` on 21 Apr 2017, the latest 1.5.x is `1.5.22.RELEASE` from 6 Aug 2019, the latest release is `2.2.2.RELEASE` from 6 Dec 2019 with the first 2.x release, `2.0.0.RELEASE` from 1 Mar 2018.

So IMHO it still does make sense to support the 1.5.x release train and maybe even go back to 1.5.0 but we should also make sure we work with the 2.x train.

Btw. to test our spring boot support with different versions of spring boot, we'd need to have an integration test maven module per spring boot version and run some shared tests within each such module. Maybe would be good to set this up at least for each minor version of spring boot (e.g. currently for 1.5.0, 2.0.0, 2.1.0 and 2.2.0)?

File: sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java [hunk 1]
```diff
@@ -58,6 +54,39 @@ public ServletContextInitializer sentryServletContextInitializer() {
         return new SentryServletContextInitializer();
     }
 
+    /**
+     * Initializes a {@link List<EventBuilderHelper>}.
+     *
+     * @return a new instance of {@link List<EventBuilderHelper>}.
+     */
+    @Bean
+    @ConditionalOnMissingBean(EventBuilderHelper.class)
```
Comment (metlos) [inline sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java]: How would this interact if the user application had similar logic to provide a different list of event builder helpers?

Comment (Patrick0308) [reply sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java]: This init method will be ignored, if spring context already has a list of builder helpers.

**Comment by Patrick0308** [comment]
@metlos Do you have some examples about testing application in different versions?

File: sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java (hunk 1 above)
Comment (Patrick0308) [reply sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java]: image

**Comment by metlos** [comment]
> @metlos Do you have some examples about testing application in different versions?

I rather meant reverting our spring dep to a version where the error would exhibit itself and write a test that makes sure that it doesn't fall over anymore.

File: sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java (hunk 1 above)
Comment (metlos) [reply sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java]: How does spring ensure the order of initialization? I.e. how is it ensured that these initializers are only used after the application-defined ones?

Comment (Patrick0308) [reply sentry-spring-boot-starter/src/main/java/io/sentry/spring/autoconfigure/SentryAutoConfiguration.java]: ok, There's a lot to say about this. In short, when initing sentry client and spring context don't has this beans, spring will search initializers which providing beans.

**Comment by Patrick0308** [comment]
Why pr can't pass CI testing. Here is the error message. 
```
Caused by: org.apache.maven.wagon.TransferFailedException: Failed to transfer file: http://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-test-support/1.5.0.RELEASE/spring-boot-test-support-1.5.0.RELEASE.pom. Return code is: 501 , ReasonPhrase:HTTPS Required.
	at org.apache.maven.wagon.providers.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:1039)
	at org.apache.maven.wagon.providers.http.AbstractHttpClientWagon.fillInputData(AbstractHttpClientWagon.java:977)
	at org.apache.maven.wagon.StreamWagon.getInputStream(StreamWagon.java:116)
	at org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:88)
	at org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61)
	at org.eclipse.aether.transport.wagon.WagonTransporter$GetTaskRunner.run(WagonTransporter.java:560)
	at org.eclipse.aether.transport.wagon.WagonTransporter.execute(WagonTransporter.java:427)
	at org.eclipse.aether.transport.wagon.WagonTransporter.get(WagonTransporter.java:404)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$GetTaskRunner.runTask(BasicRepositoryConnector.java:447)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:350)
	... 48 more
```
I found spring-boot-test-support-1.5.0.RELEASE in central maven repo.

**Comment by metlos** [comment]
Huh, I recall someone on the twitter mentioning HTTP transport is being switched off on maven central...

**Comment by metlos** [comment]
It took me a looong time to get back here but reading https://spring.io/blog/2017/01/30/spring-boot-1-5-1-released#what-happened-to-1-5-0 I assume 1.5.0 can't really be used and the minimum version of the 1.5.x should 1.5.1.

**Comment by Patrick0308** [comment]
@metlos thanks very much. What I should do next?

**Comment by bruno-garcia** [comment]
Thanks @Patrick0308 and @metlos

**Pull request merged**